#162948848 Implement Notifications

andrewhingah opened the pull request:
**What does this PR do?**
Implements email and in-app notifications

**Description of tasks to be completed**
Implement email and in-app notifications. Notifications are sent:
- to a user when that user is followed
- to followers of a user when the followed user posts new articles

**How should this be manually tested?**
Clone this repo `git clone https://github.com/andela/ah-django-unchained.git`

`cd ah-django-unchained`

`git checkout ft-user-notifications-162948848`

Get started with the app as from the README section. Follow the sample .env file supplied to obtain the right variables

Create several (at least two) users with `http://127.0.0.1:8000/api/users/`

Make a follow request on `http://127.0.0.1:8000/api/profiles/<username1>/follow` with the token header from user2

View user1 notifications on `http://127.0.0.1:8000/api/notifications/all/` using user1 token header

- in app
<img width="960" alt="screenshot 2019-02-01 at 11 19 54" src="https://user-images.githubusercontent.com/32644295/52110993-71621c00-2613-11e9-9374-01cdbe7c23a6.png">

- email
<img width="812" alt="screenshot 2019-02-01 at 11 23 42" src="https://user-images.githubusercontent.com/32644295/52111156-e5042900-2613-11e9-96db-8dcbb4994fa9.png">

Create an article using the followed user's credentials on `http://127.0.0.1:8000/api/articles/`

View the follower's notification
- in app
<img width="964" alt="screenshot 2019-02-01 at 11 27 14" src="https://user-images.githubusercontent.com/32644295/52111342-7a9fb880-2614-11e9-9a94-125000526c80.png">

- email
<img width="821" alt="screenshot 2019-02-01 at 11 28 00" src="https://user-images.githubusercontent.com/32644295/52111349-812e3000-2614-11e9-9e37-6dc863a4bcd8.png">

To unsubscribe from email notification click on the `unsubscribe` link in the email
<img width="836" alt="screenshot 2019-02-01 at 11 29 28" src="https://user-images.githubusercontent.com/32644295/52111415-b044a180-2614-11e9-8e93-0fb1a96496a9.png">

To unsubscribe from in app notifications, make a put request using `http://127.0.0.1:8000/api/notifications/unsubscribe/` and pass a token.
<img width="880" alt="screenshot 2019-02-01 at 11 30 22" src="https://user-images.githubusercontent.com/32644295/52111448-cc484300-2614-11e9-906b-2579b825a276.png">

**Relevant Pivotal Tracker Story**
#162948848

andrewhingah committed:
feat(notifications): implement user notifications

- implement app and email notifications
- users get notifications when they are followed
- users get notifications when users they follow post new articles

[Delivers #162948848]

andrewhingah committed:
feat(tests) add test for article creation in-app notification

- add test for article creation in-app notification

[Delivers #162948848]

andrewhingah committed:
feat(notifications): enable users to opt out of email notifications

- fix unsubscribe from email notifications link
- filter only subscribed users to receive notifications

[Delivers #162948848]

andrewhingah committed:
feat(notifications): implement notifications opt-in

- implement notifications opt-in for in app
- filter only subscribed users to receive notifications

[Delivers #162948848]

andrewhingah committed:
feat(notifications): test for unsubscribing to in-app notifications

- add tests for unsubscribing to in-app notifications
- add save method when notification status is changed

[Delivers #162948848]

andrewhingah committed:
feat(tests): add test for subscribing to notifications

- add more tests for subscribing back to notifications

[Deliver #162948848]

KMaina commented:
Check your requirements.txt file and remove all dependancies not used

MaggieChege commented on authors/apps/usernotifications/views.py:
```diff
@@ -0,0 +1,108 @@
+        """
+        unsubscribe from app notifications
+        """
+        self.permission_classes.append(IsAuthenticated)
```
Correct this

MaggieChege commented on authors/apps/usernotifications/views.py:
```diff
@@ -0,0 +1,108 @@
+
+class SubscribeAPIView(UpdateAPIView):
+    """
+    allow users to subscribe to nofications
```
typo

andrewhingah committed:
feat(requirements) remove unnecessary freezed dependencies

- remove unnecessary freezed dependencies

[Fixes #162948848]

andrewhingah committed:
feat(tests): add more test for response

- test for response message returned in notifications

[Delivers #162948848]

andrewhingah committed:
feat(test): test unauthenticated user cannot subscribe to notifications

- add test to check unauthenticated user cannot subscribe

[Delivers #162948848]

ignatiusukwuoma commented on authors/apps/usernotifications/handlers.py:
```diff
@@ -0,0 +1,94 @@
+    article_author = instance.author
+    # add the user's followers to the recipients list
+    followers = Friend.objects.select_related(
+        'user_from', 'user_to').filter(user_to=article_author.id).all()
```
You only need `user_from`

ignatiusukwuoma commented on authors/apps/usernotifications/handlers.py:
```diff
@@ -0,0 +1,94 @@
+        'user_from', 'user_to').filter(user_to=article_author.id).all()
+    recipients = [get_user_model().objects.get(id=u.user_from_id) for u in list(followers)]
+    for user in recipients:
+        recipient = User.objects.get(email=user)
```
Is this not the same thing that you've done on line 44?

ignatiusukwuoma commented on authors/apps/usernotifications/handlers.py:
```diff
@@ -0,0 +1,94 @@
+    for user in recipients:
+        recipient = User.objects.get(email=user)
+        if recipient.app_notification_subscription is False:
+            return
```
This should mean that if any of the recipients have `app_notification_subscription` of False, then the function should return and NOT send to anyone. Is that what you intend?

andrewhingah replied:
I intended the function to send notifications to subscribed users apart from those with `app_notification_subscription` as `False`. Let me correct that.

ignatiusukwuoma commented on authors/apps/usernotifications/handlers.py:
```diff
@@ -0,0 +1,94 @@
+    recipient = followed
+    if recipient.email_notification_subscription is False:
+        return
+    url = 'api/profiles/{}/follow/'.format(followed.username)
```
The url we send in the notification, shouldn't it be the profile url of the follower?

andrewhingah replied:
Yes it should. I'll fix that.

ignatiusukwuoma commented on authors/apps/usernotifications/handlers.py:
```diff
@@ -0,0 +1,94 @@
+        if recipient.app_notification_subscription is False:
+            return
+
+    url = "api/articles/"
```
I think we can send the link to the published article

andrewhingah replied:
Sure, though this will work when the work on `saving article as draft is merged first`.

andrewhingah replied:
I have worked on this

ignatiusukwuoma commented on authors/apps/usernotifications/urls.py:
```diff
@@ -0,0 +1,28 @@
+
+urlpatterns = [
+    path(
+        'notifications/all/',
```
No need for `all/`

ignatiusukwuoma commented on authors/apps/usernotifications/urls.py:
```diff
@@ -0,0 +1,28 @@
+        ),
+    path(
+        'notifications/unsubscribe/',
+        views.UnsubscribeAPIView.as_view(),
```
Dont Repeat Yourself, both paths use same view. Consider using re_path https://docs.djangoproject.com/en/2.1/ref/urls/#re-path

ignatiusukwuoma commented on authors/apps/usernotifications/tests/test_notifications.py:
```diff
@@ -0,0 +1,198 @@
+            response1.data['message'],
+            "You have successfully unsubscribed from app notifications")
+        self.assertEqual(response1.data["email"], True)
+        self.assertEqual(response1.data["app"], False)
```
assertTrue and assertFalse will be better here

ignatiusukwuoma commented on authors/apps/usernotifications/tests/test_notifications.py:
```diff
@@ -0,0 +1,198 @@
+            response6.data['message'],
+            "You have successfully subscribed to notifications")
+        self.assertEqual(response6.data["email"], True)
+        self.assertEqual(response6.data["app"], True)
```
ditto

andrewhingah committed:
feat(notifications): refactor all notification url, refactor article handler

- remove 'all' in get all notifications url
- refactor article handler to send notifications to subscribed recipients
- refactor tests

[Delivers #162948848]

ignatiusukwuoma commented:
Rebase branch

andrewhingah committed:
feat(notifications): if condition to send notification only when article is published

- add condition to send notification only when an article is published

[Fixes #162948848]
